source/Applications/Advanced/TransformViaArucoMarker/TransformViaArucoMarker.cpp

```diff
@@ -0,0 +1,315 @@
+            const int y2 = clamp(static_cast<int>(std::ceil(y)), 0, height - 1);
+
+            // estimate the 3d center of the marker using bilinear interpolation
+            // See https://en.wikipedia.org/wiki/Bilinear_interpolation
+            const auto q11 = points(y1, x1);
+            const auto q12 = points(y2, x1);
+            const auto q21 = points(y1, x2);
+            const auto q22 = points(y2, x2);
+
+            const float a = (x2 - x) / (x2 - x1);
+            const float b = (x - x1) / (x2 - x1);
+
+            // linear interpolation in x direction
+            // f(x,y1) = a * f(q11) + b * f(q21)
+            const cv::Point3f fxy1{ a * q11.x + b * q21.x, a * q11.y + b * q21.y, a * q11.z + b * q21.z };
+            // f(x,y2) = a * f(q12) + b * f(q22)
+            const cv::Point3f fxy2{ a * q12.x + b * q22.x, a * q12.y + b * q22.y, a * q12.z + b * q22.z };
+
+            // linear interpolation in y-axis
+            // f(x,y) = c * f(x,y1) + d * f(x,y2)
+            const float c = (y2 - y) / (y2 - y1);
+            const float d = (y - y1) / (y2 - y1);
+            markerPoints3D.emplace_back(
+                cv::Point3f{ c * fxy1.x + d * fxy2.x, c * fxy1.y + d * fxy2.y, c * fxy1.z + d * fxy2.z });
+        }
+
+        return markerPoints3D;
+    }
+
+    Zivid::Matrix4x4 transformationMatrix(const cv::Matx33d &rotationMatrix, const cv::Vec3d &translationVector)
+    {
+        auto transformMatrix = Zivid::Matrix4x4{ 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1 };
+
+        for(size_t row = 0; row < rotationMatrix.rows; row++)
+        {
+            for(size_t col = 0; col < rotationMatrix.cols; col++)
+            {
+                transformMatrix(row, col) = rotationMatrix(row, col);
+            }
+        }
+
+        transformMatrix(0, 3) = translationVector[0];
+        transformMatrix(1, 3) = translationVector[1];
+        transformMatrix(2, 3) = translationVector[2];
+
+        return transformMatrix;
+    }
+
+    void displayMarkerPose(const Zivid::PointCloud &pointCloud,
+                           const std::vector<cv::Point2f> &corners2D,
+                           const std::vector<cv::Point2f> &center2D)
+    {
+        const auto bgr = pointCloudToCvBGR(pointCloud);
+
+        auto poseCenter = center2D[0];
+        auto xAxisPoint = cv::Point2f(((corners2D[3].x + corners2D[0].x) / 2), ((corners2D[3].y + corners2D[0].y) / 2));
+        auto yAxisPoint = cv::Point2f(((corners2D[2].x + corners2D[3].x) / 2), ((corners2D[2].y + corners2D[3].y) / 2));
+
+        cv::circle(bgr, poseCenter, 2, cv::Scalar(255, 0, 0), 2);
+        cv::arrowedLine(bgr, poseCenter, xAxisPoint, cv::Scalar(0, 0, 255), 2);
+        cv::arrowedLine(bgr, poseCenter, yAxisPoint, cv::Scalar(0, 255, 0), 2);
+
+        cv::imshow("Pose of aruco marker", bgr);
+        cv::waitKey(0);
+    }
+
+    Zivid::Matrix4x4 getMarkerPose(const Zivid::PointCloud &pointCloud, const aruconano::Marker &markers)
```

Comment:
rename to getAndDisplayMarkerPose if you display in this function. Alternative is to take displayMarkerPose() as a separate function and call it outside this one.

Reply:
better to call it getAndDisplayMarkerPose, since it requires 2D corner and centers to display the pose. These are not returned from the function